services/src/main/java/org/keycloak/protocol/oidc/endpoints/UserInfoEndpoint.java

```diff
@@ -251,6 +254,18 @@ private Response issueUserInfo() {
             }
         }
 
+        if (Profile.isFeatureEnabled(Profile.Feature.DPOP)) {
+            if (OIDCAdvancedConfigWrapper.fromClientModel(clientModel).isUseDPoP() || request.getHttpHeaders().getHeaderString(DPoPUtil.DPOP_HTTP_HEADER) != null) {
```

Comment:
This means that if client doesn't require DPoP, but still DPoP was used during authentication, the validation won't be done in case that `DPoP` header is not present in UserInfo endpoint.

To me, it would makes sense to require validate DPoP always if it was used during authentication? In case that accessToken contains thumbprint, it means that DPoP was used during authentication and in that case, it might be always verified? So I can think about something like this:
```
            if (OIDCAdvancedConfigWrapper.fromClientModel(clientModel).isUseDPoP() || accessTokenHasDPoPThumbprint()) {
```
WDYT? Can that work?

Reply:
Yes, you are right. The current PR bypasses the DPoP verification when a client sends DPoP bound access token without DPoP proof header. I will fix the PR as you suggested.

Reply:
Actually, I think that if we address https://github.com/keycloak/keycloak/issues/21919, the token type should be set as `DPoP`. So in that case, the check can be even simpler as:
```
if (OIDCAdvancedConfigWrapper.fromClientModel(clientModel).isUseDPoP() || DPoPUtil.DPOP_TOKEN_TYPE.equals(token.getType())) {
```
or something like that.

Reply:
Yes, I used `DPoPUtil.DPOP_TOKEN_TYPE.equals(token.getType())`.